### DISCUSSION TITLE
what's the purpose of "quality" when encoding LOSSLESS wepb?

### DISCUSSION THREAD

**Discussion by shula** [ImageMagick/ImageMagick#7763] [Help]
### ImageMagick version

7.1.1-39 Q16-HDRI

### Operating system, version and so on

linux (various)

### Description

When encoding an image to WEBP, it's possible to set different quality levels.  
(Gimp, OTOH, disables 'quality' if "lossless" is enabled).

The image size is significantly different, yet the contents looks identical !
(compared via naked eye).

1) What does "quality" mean when WEBP is lossless?
2) if the output is the same, why the size difference?

examples:
```
magick a.png -quality 5 -define webp:lossless=true a5.webp
magick a.png -quality 99 -define webp:lossless=true a99.webp
``` 

## size in bytes; all outputs are lossless:
```
 47393  a.png             # original
 19018  a-05.webp     # quality 5
 23566  a-99.webp     # quality 99
 19948  a-50.webp     # quality 50
 25956  a-gimp.webp     # quality N/A
```

![image](https://github.com/user-attachments/assets/862cc49c-deec-48af-8c42-12e0fadc6394)

**Comment by fmw42** [comment]
Try putting -quality after reading the input and before the output.  IM 7 syntax is different from IM 6 and different from Unix.

See also https://imagemagick.org/script/webp.php

**Reply by shula** [reply]
isn't it what i did already?
`magick a.png -quality 5 -define webp:lossless=true a5.webp`
"a.png" is the input, so "quality is after the input, and a5.webp, which is the output.
In any case, it doesn't address my question.

**Reply by fmw42** [reply]
Sorry I misread your post. You have it correct. What is your version of libwebp? And is that current?

**Reply by shula** [reply]
libwebp 1.4.0-3

**Comment by snibgo** [comment]
> The image size is significantly different, yet the contents looks identical !
(compared via naked eye).

They look identical, but they are not quite.

If the input is only 8 bits/channel/pixel, then "`-quality 100 -define webp:lossless=true`" gives an identical output. Quality numbers less than 100 give smaller files that are more different to the input. In my tests on ordinary photos, the difference is always less than 1% RMSE, and the difference is not visible.

If we remove "`-define webp:lossless=true`" then smaller quality numbers give much smaller files with larger RMSE numbers eg 4% and the difference is visibly obvious.
